### PR TITLE
Add docs for per key inference

### PR DESCRIPTION
**Pull request by damccorm**
This should not be merged until release 2.51 is finalized.

✨ [RENDERED](http://apache-beam-website-pull-requests.storage.googleapis.com/28243/documentation/sdks/python-machine-learning/index.html#use-a-keyed-modelhandler)

Part of #27628

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (damccorm) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:236]: TODO - this still needs to talk about memory pitfalls/memory management

**Comment by damccorm** [comment]
R: @tvalentyn

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (tvalentyn) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:253]: > load at most 2 models per worker 

Users might perceive it as a guarantee, and come to us if/when they see a single OOM error.  If this cannot be guaranteed, we can phrase that the upper  ceiling is enforced as best effort. Or mention that there may be some delay between when the model is unloaded and the memory is freed.

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (tvalentyn) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:255]: Since worker is such an overloaded term, users might be confused. In Dataflow context, `worker` often refers to the  VM, as in  `max_num_workers` pipeline option.

Here by 'worker' you refer to SDK worker process. In  https://cloud.google.com/dataflow/docs/guides/troubleshoot-oom#sdk-process-memory we call these  'Apache Beam SDK process`. How about you use that or `SDK worker process` instead of `worker`?

**Review by tvalentyn — COMMENTED** [review]
Very nice, thanks, @damccorm ! Consider getting a review from TW folks before merging as well.

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (damccorm) [reply website/www/site/content/en/documentation/sdks/python-machine-learning.md]: I haven't yet, but its a good idea I can follow up with

Comment (damccorm) [reply website/www/site/content/en/documentation/sdks/python-machine-learning.md]: I mentioned that there's some delay, let me know if you think the wording is ok

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (damccorm) [reply website/www/site/content/en/documentation/sdks/python-machine-learning.md]: Updated references to `SDK worker process`

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (damccorm) [reply website/www/site/content/en/documentation/sdks/python-machine-learning.md:218]: Yes, I thought I had already done that. Updated

**Comment by damccorm** [comment]
retest this please

**Review by tvalentyn — APPROVED** [review]

**Comment by damccorm** [comment]
Run Portable_Python PreCommit

**Comment by damccorm** [comment]
Run PythonDocker PreCommit

**Comment by damccorm** [comment]
Run PythonDocs PreCommit

**Comment by damccorm** [comment]
Run Python_Coverage PreCommit

**Comment by damccorm** [comment]
Run Python_Examples PreCommit

**Comment by damccorm** [comment]
Run Python_Integration PreCommit

**Comment by damccorm** [comment]
Run Python_PVR_Flink PreCommit

**Comment by damccorm** [comment]
Run Python_Runners PreCommit

**Comment by damccorm** [comment]
Run Python_Transforms PreCommit

**Comment by damccorm** [comment]
Run Website PreCommit

**Comment by damccorm** [comment]
Run Website_Stage_GCS PreCommit

**Comment by damccorm** [comment]
R: @rszper

File: sdks/python/apache_beam/ml/inference/base.py
Comment (rszper) [inline sdks/python/apache_beam/ml/inference/base.py:485]:
```suggestion
    Loading multiple models at the same time can increase the risk of an out of
```

File: sdks/python/apache_beam/ml/inference/base.py
Comment (rszper) [inline sdks/python/apache_beam/ml/inference/base.py:486]:
```suggestion
    memory (OOM) exception. To avoid this issue, use the parameter
```

File: sdks/python/apache_beam/ml/inference/base.py
Comment (rszper) [inline sdks/python/apache_beam/ml/inference/base.py:487]:
```suggestion
    `max_models_per_worker_hint` to limit the number of models that are loaded at the same time.
```

File: sdks/python/apache_beam/ml/inference/base.py
Comment (rszper) [inline sdks/python/apache_beam/ml/inference/base.py:488]:
```suggestion
    For more information about memory management, see
```

File: sdks/python/apache_beam/ml/inference/base.py
Comment (rszper) [inline sdks/python/apache_beam/ml/inference/base.py:489]:
```suggestion
    [Use a keyed `ModelHandler`](https://beam.apache.org/documentation/sdks/python-machine-learning/#use-a-keyed-modelhandler).
```

File: sdks/python/apache_beam/ml/inference/base.py
Comment (rszper) [inline sdks/python/apache_beam/ml/inference/base.py:500]:
```suggestion
        in memory with some buffer. For more information about memory management,
```

File: sdks/python/apache_beam/ml/inference/base.py
Comment (rszper) [inline sdks/python/apache_beam/ml/inference/base.py:502]:
```suggestion
        [Use a keyed `ModelHandler`](https://beam.apache.org/documentation/sdks/python-machine-learning/#use-a-keyed-modelhandler).
```

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:235]:
```suggestion
The previous example loads a model by using `config1`. That model is then used for inference for all examples associated
```

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:236]:
```suggestion
with `key1`. It then loads a model by using `config2`. That model is used for all examples associated with `key2` and `key3`.
```

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:238]:
```suggestion
Loading multiple models at the same times increases the risk of out of memory (OOM) errors. By default, `KeyedModelHandler` doesn't
```

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:239]:
```suggestion
limit the number of models loaded into memory at the same time. If the models don't all fit into memory,
```

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:240]:
```suggestion
your pipeline will likely fail with an out of memory error. To avoid this issue, provide a hint about the
```

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:241]:
```suggestion
maximum number of models that can be loaded at the same time.
```

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:253]:
```suggestion
The previous example loads at most two models per SDK worker process at any given time. It unloads models that aren't
```

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:254]:
```suggestion
currently being used. Runners that have multiple SDK worker processes on a given machine load at most
```

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:255]:
```suggestion
`max_models_per_worker_hint*<num worker processes>` models onto the machine. Leave enough space for the models
```

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:256]:
```suggestion
and any additional memory needs from other transforms. Because there might be a delay between when a model is offloaded and when the
```

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:257]:
```suggestion
memory is released, it is recommended that you leave additional buffer.
```

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:259]:
```suggestion
**Note**: Having many models but a small `max_models_per_worker_hint` can lead to _memory thrashing_, where
```

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:261]:
```suggestion
of memory thrashing, if you're using a distributed runner, insert a
```

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:263]:
```suggestion
inference step. This step reduces thrashing by ensuring that elements with the same key and model are
```

File: website/www/site/content/en/documentation/sdks/python-machine-learning.md
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-machine-learning.md:264]:
```suggestion
collocated on the same worker.
```

**Pull request merged**
